### PR TITLE
Exclude withheld ops in agent activity

### PR DESCRIPTION
**Pull request by ThetaSinner**
### Summary

Contributes to https://github.com/holochain/holochain/issues/4268

### TODO:
- [ ] CHANGELOGs updated with appropriate info
- [ ] All code changes are reflected in docs, including module-level docs

### CODE REVIEW

File: crates/holochain_cascade/src/authority/get_agent_activity_query/hashes.rs [hunk 1]
```diff
@@ -53,6 +53,7 @@ impl Query for GetAgentActivityHashesQuery {
                 -- is an action authored by this agent
                 Action.author = :author
                 AND DhtOp.type = :chain_op_type
+                AND DhtOp.withhold_publish IS NULL
```
Comment (ThetaSinner) [inline crates/holochain_cascade/src/authority/get_agent_activity_query/hashes.rs:56]: This field is only ever set in the authored database. So this is only a problem if the author ends up answering an activity query for themselves and the query falls back to the authored database. Still, it seems like a hole worth closing.

File: crates/holochain_cascade/src/authority/test.rs [hunk 2]
```diff
@@ -295,3 +294,108 @@ async fn get_agent_activity() {
...
+    )
+    .await
+    .unwrap();
+
+    let hashes_without_head = match &td.valid_hashes {
+        ChainItems::Hashes(hashes) => hashes.iter().take(hashes.len() - 1).cloned().collect(),
+        _ => unreachable!(),
+    };
+    let chain_head = match &td.valid_hashes {
+        ChainItems::Hashes(hashes) => hashes.iter().nth_back(1).cloned().unwrap(),
```
Comment (jost-s) [inline crates/holochain_cascade/src/authority/test.rs:335]: ?
```suggestion
        ChainItems::Hashes(hashes) => hashes.iter().first().cloned().unwrap(),
```

**Review by jost-s — APPROVED** [review]
Yes, worth closing!

File: crates/holochain_cascade/src/authority/test.rs (hunk 2 above)
Comment (ThetaSinner) [reply crates/holochain_cascade/src/authority/test.rs:335]: In a list of 50 items, with 0 based indexing - `first_back(1)` is index 48, where `first` is index 0

I know it looks weird, but clippy shouted at me when I tried to do this a way that looked natural to me :D

File: crates/holochain_cascade/src/authority/get_agent_activity_query/hashes.rs (hunk 1 above)
Comment (maackle) [reply crates/holochain_cascade/src/authority/get_agent_activity_query/hashes.rs:56]: note, https://github.com/holochain/holochain/pull/4295 may help with this kind of thing, ensuring that a query only runs over one of the ops databases

File: crates/holochain_cascade/src/authority/test.rs [hunk 3]
```diff
@@ -295,3 +294,108 @@ async fn get_agent_activity() {
 
     pretty_assertions::assert_eq!(result, expected);
 }
+
+#[tokio::test(flavor = "multi_thread")]
+async fn get_agent_activity_respects_withhold_publish() {
+    holochain_trace::test_run();
+    let db = test_dht_db();
```
Comment (maackle) [inline crates/holochain_cascade/src/authority/test.rs:301]: Should this be an authored db? Not that it currently matters, but semantically it's true, right?

File: crates/holochain_cascade/src/authority/test.rs (hunk 2 above)
Comment (maackle) [reply crates/holochain_cascade/src/authority/test.rs:335]: It's hard to understand why you'd want the second-to-last item and not the last item? Maybe the variable should be called something other than `chain_head`?

File: crates/holochain_cascade/src/authority/get_agent_activity_query/hashes.rs (hunk 1 above)
Comment (ThetaSinner) [reply crates/holochain_cascade/src/authority/get_agent_activity_query/hashes.rs:56]: That would certainly help with queries that are only supposed to run on one kind of database. This query lives in the cascade but it's actually the pattern from the `holochain_state` crate, which provides queries that can be run against cache/dht/authored databases.

Maybe it's not even valid that we're answering activity queries from our authored data? If we're acting as an authority for ourselves then we should probably only return data that has been placed in the DHT database. 

This is a bit of a general question for me - should we be able to describe this in general to the cascade? Some queries are valid to be run against any database and it's useful that if the DHT database is being rebuilt for some reason, that we can query our authored data. But there are maybe some cases like this where we should maybe not do it

Comment (ThetaSinner) [reply crates/holochain_cascade/src/authority/get_agent_activity_query/hashes.rs:56]: Never mind, this query is actually only called with a DHT database... So we're already doing that. Maybe this change isn't valid then and there's a different explanation for why agent activity queries to the author are resulting in countersigning sessions completing in Jost's test...

I'm going to put this back into draft for now and see if we can figure out why that's happening

**Pull request closed without merging**